src/Microsoft.AspNetCore.Mvc.Core/Controllers/ControllerFeatureProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+
+namespace Microsoft.AspNetCore.Mvc.Controllers
+{
+    /// <summary>
+    /// Discovers controllers from a list of <see cref="ApplicationPart"/> instances.
+    /// </summary>
+    public class ControllerFeatureProvider : IApplicationFeatureProvider<ControllerFeature>
+    {
+        private const string ControllerTypeNameSuffix = "Controller";
+
+        /// <inheritdoc />
+        public void PopulateFeature(
+            IEnumerable<ApplicationPart> parts,
+            ControllerFeature feature)
+        {
+            foreach (var part in parts.OfType<IApplicationPartTypeProvider>())
+            {
+                foreach (var type in part.Types)
+                {
+                    if (IsController(type) && !feature.Controllers.Contains(type))
```

Comment:
The guard is there (an I believe it's necessary) in case you have multiple IApplicationFeatureProviders or you invoke Populate feature on the part manager with an already populated feature. I want to avoid duplicates as that will cause trouble at a later point in the pipeline.

I cared to the extent of putting IsController before in the clause to avoid unnecessary lookups, but this code is going to be called once during startup and we don't have data to suggest that it would be relevant in terms of time, so I'll take some measurements to make sure. What is the worst n we would consider in this scenario?